WeakAuras/Prototypes.lua

```diff
@@ -1135,20 +1135,35 @@ local function AddUnitChangeInternalEvents(unit, t)
   end
 end
 
+local function AddUnitEventForEvents(unit, result, event)
+  if not WeakAuras.baseUnitId[unit] then
+    tinsert(result, event)
+  end
+end
+
+local function AddUnitEventForUnit_events(unit, result, event)
+  if WeakAuras.baseUnitId[unit] then
+    result[unit] = result[unit] or {}
+    tinsert(result[unit], event)
+  end
+end
+
 WeakAuras.event_prototypes = {
   ["Unit Characteristics"] = {
     type = "status",
     events = function(trigger)
```

Comment:
It might be good to instead of introducing a separate unit_events function, merge the unit_events function into events, but make it return two results, that is: 
1) The events
2) The unit events

Then, you can write a simple AddUnitEventForEvents(result, unit_result, unit, event) function which adds the event to the right place.

Reply:
that would remove a lot of duplicate code indeed

Reply:
before going further, would that structure work or you have something better in mind ?
```Lua
local function AddUnitEventForEvents(result, unit, event)
  if not WeakAuras.baseUnitId[unit] then
    if not result.events then
      result.events = {}
    end
    tinsert(result.events, event)
  else
    if not result.unit_events then
      result.unit_events = {}
    end
    if not result.unit_events[unit] then
      result.unit_events[unit] = {}
    end
    tinsert(result.unit_events[unit], event)
  end
end
```